gsy_framework/data_classes.py

```diff
@@ -324,44 +381,43 @@ def from_json(trade_bid_offer_info: str) -> "TradeBidOfferInfo":
 
 class Trade:
     """Trade class."""
-    def __init__(self, id: str, creation_time: DateTime, offer_bid: Union[Offer, Bid],
-                 seller: str, buyer: str,
+    def __init__(self, id: str, creation_time: DateTime,
+                 seller: TraderDetails, buyer: TraderDetails,
                  traded_energy: float, trade_price: float,
+                 offer: Offer = None,
+                 bid: Bid = None,
                  residual: Optional[Union[Offer, Bid]] = None,
                  already_tracked: bool = False,
                  offer_bid_trade_info: Optional[TradeBidOfferInfo] = None,
-                 seller_origin: Optional[str] = None, buyer_origin: Optional[str] = None,
-                 fee_price: Optional[float] = None, seller_origin_id: Optional[str] = None,
-                 buyer_origin_id: Optional[str] = None, seller_id: Optional[str] = None,
-                 buyer_id: Optional[str] = None, time_slot: Optional[DateTime] = None,
+                 fee_price: Optional[float] = None, time_slot: Optional[DateTime] = None,
                  matching_requirements: Optional[Dict] = None):
 
         self.id = str(id)
         self.creation_time = creation_time
         self.time_slot = time_slot  # market slot of creation
-        self.offer_bid = offer_bid
-        self.seller = seller
-        self.buyer = buyer
         self.traded_energy = traded_energy
         self.trade_price = trade_price
         self.residual = residual
         self.already_tracked = already_tracked
         self.offer_bid_trade_info = offer_bid_trade_info
-        self.seller_origin = seller_origin
-        self.buyer_origin = buyer_origin
         self.fee_price = fee_price
-        self.seller_origin_id = seller_origin_id
-        self.buyer_origin_id = buyer_origin_id
-        self.seller_id = seller_id
-        self.buyer_id = buyer_id
+        self.seller = seller
+        self.buyer = buyer
         self.matching_requirements = matching_requirements
+        self.match_details = {
+            "offer": offer, "bid": bid
+        }
```

Comment:
I wonder now: We really only need the `bid.id` and the `offer.id` for the orderbook, right? All the other information about the seller and the buyer is already saved in the trade object. And we are not interested in the offers and bids energy and price (only for debugging I guess). 
Do you thin we could go the following way in a future enhancement?
```suggestion
        self.match_details = {
            "offer_id": offer.id, "bid_id": bid.id
        }
```

Reply:
I would very much like it if we only needed to maintain the ids of the bid and the offer. Unfortunately though, the B2B product requires a trade to list more information about the bid and the offer that it originated from (including the total energy of the bid and the offer, the price of the bid and offer amongst others). So, despite my reluctance to add the full bid and offer object here, I had to include it. 
As to why there are separate parameters for the buyer / seller in addition to the bid / offer (because we could potentially only use the bid / offer which already contain the buyer and the seller information), that was also a hard decision that I had to make, because the offer or bid can be None in certain cases (chain trades). In this case, the buyer / seller would have to be populated separately. I was thinking to create an "empty Order" structure, that would contain only buyer / seller information without the other order info, in order to use instead of None in the bid / offer key in such cases, which would remove the duplication of information, however this would make the structure of the class more complex than it is ATM. Therefore I opted to accept the duplication for ease of use of the class.